node/core/dispute-participation/src/lib.rs

```diff
@@ -0,0 +1,370 @@
+{
+	let mut state = State { recent_block: None };
+
+	loop {
+		match ctx.recv().await {
+			Err(_) => return,
+			Ok(FromOverseer::Signal(OverseerSignal::Conclude)) => {
+				tracing::info!(target: LOG_TARGET, "Received `Conclude` signal, exiting");
+				return;
+			}
+			Ok(FromOverseer::Signal(OverseerSignal::BlockFinalized(_, _))) => {}
+			Ok(FromOverseer::Signal(OverseerSignal::ActiveLeaves(update))) => {
+				update_state(&mut state, update);
+			}
+			Ok(FromOverseer::Communication { msg }) => {
+				if let Err(err) = handle_incoming(&mut ctx, &mut state, msg).await {
+					err.trace();
+					if let Error::Subsystem(SubsystemError::Context(_)) = err {
+						return;
+					}
+				}
+			}
+		}
+	}
+}
+
+fn update_state(state: &mut State, update: ActiveLeavesUpdate) {
+	for active in update.activated {
+		if state.recent_block.map_or(true, |s| active.number > s.0) {
+			state.recent_block = Some((active.number, active.hash));
+		}
+	}
+}
+
+async fn handle_incoming(
+	ctx: &mut impl SubsystemContext,
+	state: &mut State,
+	message: DisputeParticipationMessage,
+) -> Result<(), Error> {
+	match message {
+		DisputeParticipationMessage::Participate {
+			candidate_hash,
+			candidate_receipt,
+			session,
+			n_validators,
+		} => {
+			if let Some((_, block_hash)) = state.recent_block {
+				participate(
+					ctx,
+					block_hash,
+					candidate_hash,
+					candidate_receipt,
+					session,
+					n_validators,
+				)
+				.await
+			} else {
+				return Err(ParticipationError::MissingRecentBlockState.into());
+			}
+		}
+	}
+}
+
+async fn participate(
+	ctx: &mut impl SubsystemContext,
+	block_hash: Hash,
+	candidate_hash: CandidateHash,
+	candidate_receipt: CandidateReceipt,
+	session: SessionIndex,
+	n_validators: u32,
+) -> Result<(), Error> {
+	let (recover_available_data_tx, recover_available_data_rx) = oneshot::channel();
+	let (code_tx, code_rx) = oneshot::channel();
+	let (store_available_data_tx, store_available_data_rx) = oneshot::channel();
+	let (validation_tx, validation_rx) = oneshot::channel();
+
+	// in order to validate a candidate we need to start by recovering the
+	// available data
+	ctx.send_message(
+		AvailabilityRecoveryMessage::RecoverAvailableData(
+			candidate_receipt.clone(),
+			session,
+			None,
+			recover_available_data_tx,
+		)
+		.into(),
+	)
+	.await;
+
+	let available_data = match recover_available_data_rx.await? {
+		Ok(data) => data,
+		Err(RecoveryError::Invalid) => {
+			// the available data was recovered but it is invalid, therefore we'll
+			// vote negatively for the candidate dispute
+			cast_invalid_vote(ctx, candidate_hash, candidate_receipt, session).await;
+			return Ok(());
+		}
+		Err(RecoveryError::Unavailable) => {
+			return Err(ParticipationError::MissingAvailableData.into());
+		}
+	};
+
+	// we also need to fetch the validation code which we can reference by its
+	// hash as taken from the candidate descriptor
+	ctx.send_message(
+		RuntimeApiMessage::Request(
+			block_hash,
```

Comment:
is this the only place we use `block_hash`? why do we need it again? runtime api uses it only for caching